sql-plugin/src/main/scala/com/nvidia/spark/rapids/conditionalExpressions.scala

```diff
@@ -301,12 +306,139 @@ case class GpuCaseWhen(
+                }
+                withResource(thenValues) { _ =>
+                  withResource(gather(whenBoolAdjusted.getBase, thenValues)) { gather =>
+                    currentValue = Some(currentValue match {
+                      case Some(v) =>
+                        withResource(v) { _ =>
+                          GpuColumnVector.from(whenBoolAdjusted.getBase.ifElse(gather, v.getBase),
+                            dataType)
+                        }
+                      case _ =>
+                        GpuColumnVector.from(gather.incRefCount(), dataType)
+                    })
+                  }
+
+                  cumulativePred = Some(cumulativePred match {
+                    case Some(prev) =>
+                      withResource(prev) { _ =>
+                        GpuColumnVector.from(whenBool.getBase.or(prev.getBase),
+                          DataTypes.BooleanType)
+                      }
+                    case _ =>
+                      whenBoolAdjusted.incRefCount()
+                  })
+
+                  if (isAllTrue(cumulativePred.get)) {
+                    // no need to process any more branches or the else condition
+                    return currentValue.get.incRefCount()
+                  }
+                }
+              }
+            }
+        }
+
+        // invert the cumulative predicate to get the ELSE predicate
+        withResource(cumulativePred.get.getBase.not()) { elsePredicate =>
+          // replace null predicates with true (because this is an inverted predicate)
+          val elsePredNoNulls = withResource(Scalar.fromBool(true)) { t =>
+            GpuColumnVector.from(elsePredicate.replaceNulls(t), DataTypes.BooleanType)
+          }
+          withResource(elsePredNoNulls) { _ =>
+            elseValue match {
+              case Some(expr) =>
+                if (isAllTrue(elsePredNoNulls)) {
+                  GpuExpressionsUtils.columnarEvalToColumn(expr, batch)
+                } else {
+                  val elseValues = withResource(
+                      filterBatch(tbl, elsePredNoNulls.getBase, colTypes)) {
+                    elseBatch => GpuExpressionsUtils.columnarEvalToColumn(expr, elseBatch)
+                  }
+                  withResource(elseValues) { _ =>
+                    withResource(gather(elsePredNoNulls.getBase, elseValues)) { gather =>
+                      GpuColumnVector.from(elsePredNoNulls.getBase.ifElse(
+                        gather, currentValue.get.getBase), dataType)
+                    }
+                  }
+                }
+
+              case None =>
+                // if there is no ELSE condition then we return NULL for any rows not matched by
+                // previous branches
+                withResource(GpuScalar.from(null, dataType)) { nullScalar =>
+                  if (isAllTrue(elsePredNoNulls) && elsePredNoNulls.getRowCount <= Int.MaxValue) {
```

Comment:
I don't understand the reason for the `getRowCount` check here or how it could ever evaluate to false.

Reply:
My rationale here was that `getRowCount` returns a `Long` and we need to pass it to `GpuColumnVector.from` which requires `Int`. If in the future we supported row counts > `Int.MaxValue` then the call to `toInt` could be unsafe?

Reply:
If we support row counts > `Int.MaxValue` in the future then `GpuColumnVector.from` would also need to support the same.